subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationRole.java

```diff
@@ -0,0 +1,192 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.configurations;
+
+import org.gradle.internal.deprecation.DeprecationLogger;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Defines how a {@link org.gradle.api.artifacts.Configuration} is intended to be used.
+ *
+ * Standard roles are defined in {@link ConfigurationRoles}.
+ *
+ * @since 8.0
+ */
+public interface ConfigurationRole {
+    /**
+     * Returns a human-readable name for this role.
+     */
+    String getName();
+
+    boolean isConsumable();
+    boolean isResolvable();
+    boolean isDeclarableAgainst();
+    boolean isConsumptionDeprecated();
+    boolean isResolutionDeprecated();
+    boolean isDeclarationAgainstDeprecated();
+
+    /**
+     * Obtains a human-readable summary of the usage allowed by the given role.
+     */
+    default String describe() {
+        return RoleDescriber.describeRole(this);
+    }
+
+    /**
+     * Attempts to locate a pre-defined role allowing the given usage in {@link ConfigurationRoles} and return it;
+     * if such a roles does not exist, creates a custom anonymous implementation and returns it instead.
+     *
+     * @param name the name to use when creating a role that wasn't found
+     * @param consumable whether this role is consumable
+     * @param resolvable whether this role is resolvable
+     * @param declarableAgainst whether this role is declarable against
+     * @param consumptionDeprecated whether this role is deprecated for consumption
+     * @param resolutionDeprecated whether this role is deprecated for resolution
+     * @param declarationAgainstDeprecated whether this role is deprecated for declaration against
+     * @param description a custom description to use when creating a role that wasn't found; or {@code null} to use the default description generated by {@link RoleDescriber}
+     * @return a role with matching usage characteristics
+     */
+    static ConfigurationRole forUsage(String name, boolean consumable, boolean resolvable, boolean declarableAgainst, boolean consumptionDeprecated, boolean resolutionDeprecated, boolean declarationAgainstDeprecated, @Nullable String description) {
+        if ((!consumable && consumptionDeprecated) || (!resolvable && resolutionDeprecated) || (!declarableAgainst && declarationAgainstDeprecated)) {
+            throw new IllegalArgumentException("Cannot create a role that deprecates a usage that is not allowed");
+        }
+
+        ConfigurationRole result = ConfigurationRoles.byUsage(consumable, resolvable, declarableAgainst, consumptionDeprecated, resolutionDeprecated, declarationAgainstDeprecated)
+                .map(ConfigurationRole.class::cast)
+                .orElse(new ConfigurationRole() {
```

Comment:
Creating an implementation of `ConfigurationRole` here and also in `ConfigurationRoles` seems redundant. Instead of a `ConfigurationRole` interface and a `ConfigurationRoles` enum, what if we did a `ConfigurationRole` interface with static constants, each of which are instances of a `DefaultConfigurationRole`? This way, we can avoid multiple implementations of `ConfigurationRole` and would also make it easier to write adhoc roles if we need to.

Reply:
I don't see why @jvandort's solution would result in larger memory usage -- the idea was specifically to use `static` constants to re-use the shared values. I think that would be a better setup than having large anonymous classes.

Reply:
As `static` constants it would begin with only # instances = # roles, yes.  But after a few trips to and from the configuration cache, wouldn't each configuration now have a distinct role instance after being deserialized?  I haven't looked into how that works, I just figured the one enum value per JVM guarantee is ironclad.  

If this isn't a factor, I can change it.  I don't really understand the concern about the anonymous classes or multiple implementations though, they are each very simple and I didn't think they were particularly large or complex.

Reply:
We can configure the config cache to not do that.

Personally, I think that any class that is inline with some code ought to not be more than 5-10 lines long, or it interrupts my context for the value. At over 40 lines, I think this should at least be made into a separate `static final` inner class, if the solution of using a single type is too much.

Reply:
The readability argument is a good one.  I'll extract a named class to get this out of the flow when reading.

I'm still not sold on the need for a `DefaultConfigurationRole` just to avoid multiple implementations though, even if there aren't any CC considerations.  The only reason these implementations are long is the natural verbosity of Java - they're super simple to write and understand.  A set of predefined roles makes me favor an enum which makes the intent clearer without any additional explanation, vs. constants in the interface which are marginally harder to discover.